homeassistant/components/upnp/__init__.py

```diff
@@ -0,0 +1,163 @@
+"""
+Will open a port in your router for Home Assistant and provide statistics.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/upnp/
+"""
+# pylint: disable=invalid-name
+import asyncio
+from ipaddress import ip_address
+
+import aiohttp
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import dispatcher
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.components.discovery import DOMAIN as DISCOVERY_DOMAIN
+
+from .const import (
+    CONF_ENABLE_PORT_MAPPING, CONF_ENABLE_SENSORS,
+    CONF_HASS, CONF_LOCAL_IP, CONF_PORTS,
+    CONF_UDN, CONF_SSDP_DESCRIPTION,
+)
+from .const import DOMAIN
+from .const import LOGGER as _LOGGER
+from .config_flow import ensure_domain_data
+from .device import Device
+
+
+REQUIREMENTS = ['async-upnp-client==0.12.4']
+DEPENDENCIES = ['http']
+
+NOTIFICATION_ID = 'upnp_notification'
+NOTIFICATION_TITLE = 'UPnP/IGD Setup'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_ENABLE_PORT_MAPPING, default=False): cv.boolean,
+        vol.Optional(CONF_ENABLE_SENSORS, default=True): cv.boolean,
+        vol.Optional(CONF_LOCAL_IP): vol.All(ip_address, cv.string),
+        vol.Optional(CONF_PORTS):
+            vol.Schema({
+                vol.Any(CONF_HASS, cv.positive_int):
+                    vol.Any(CONF_HASS, cv.positive_int)
+            })
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def _substitute_hass_ports(ports, hass_port):
+    # substitute 'hass' for hass_port, both sides
+    if CONF_HASS in ports:
+        ports[hass_port] = ports[CONF_HASS]
+        del ports[CONF_HASS]
+    for port in ports:
+        if ports[port] == CONF_HASS:
+            ports[port] = hass_port
+
+
+# config
+async def async_setup(hass: HomeAssistantType, config: ConfigType):
+    """Register a port mapping for Home Assistant via UPnP."""
+    ensure_domain_data(hass)
+
+    # ensure sane config
+    if DOMAIN not in config:
+        return True
+
+    if DISCOVERY_DOMAIN not in config:
+        _LOGGER.warning('UPNP needs discovery, please enable it')
+        return False
+
+    # overridden local ip
+    upnp_config = config[DOMAIN]
+    if CONF_LOCAL_IP in upnp_config:
+        hass.data[DOMAIN]['local_ip'] = upnp_config[CONF_LOCAL_IP]
+
+    # determine ports
+    ports = {CONF_HASS: CONF_HASS}  # default, port_mapping disabled by default
+    if CONF_PORTS in upnp_config:
+        # copy from config
+        ports = upnp_config[CONF_PORTS]
```

Comment:
Since we will modify ports in `_substitute_hass_ports` we probably should make a copy here.